### DISCUSSION TITLE
Initializing Scalar Field with User-Defined Function

### DISCUSSION THREAD

**Discussion by alexandre-barbosa** [zwicker-group/py-pde#428] [Q&A] [answered]
Hello,

I am using py-pde to solve a system of coupled PDE for complex (scalar) fields, where the evolution rate depends on solutions of the Helmholtz equation; which in spherical coordinates are spherical Bessel functions.  

As a minimal (not-)working example, trying to run

```
from pde import PDEBase, ScalarField, SphericalSymGrid
from scipy.special import spherical_jn
  
 class MyPDE(PDEBase):
  
      def __init__(self, f=None):
          super().__init__()
          self.f = ScalarField.from_expression(grid, f"bessel(r)", user_funcs={"bessel": lambda r: spherical_jn(1, r)})
          
      def evolution_rate(self, state, t=0):
          S = state 
          
          dS_dt = S*self.f
  
          return dS_dt

  
 grid = SphericalSymGrid(radius=(0, 1), shape=256)
  
 eq = MyPDE()
```

results in the error message

`from_expression() got an unexpected keyword argument 'user_funcs'`

I have looked into the documentation, other GitHub discussion posts and attempted several variations of this, but I always get this error. I would greatly appreciate any help in understanding this error message, or an alternative way to implement this?

Thank you so much in advance!

Alexandre

**Comment by david-zwicker** [comment]
This is unexpected. What version of `py-pde` are you running? Have you tried updating to the latest version?

**Reply by alexandre-barbosa** [reply, marked as answer]
Thank you so much for the fast reply! I am running the latest version, 0.31.0. It seems that the issue is not with py-pde, however.

I can run the code by running it from the command line, but it yields the error message above when trying to run it from Spyder. I don't understand what could cause this to happen, and I don't think it's worth it to spend time trying to do so. I am  just happy enough that there is a workaround; in case the issue is not just with my particular Python installation but with Spyder, I am sharing it here.

Again, thank you for the amazing package, it is incredibly intuitive to use!

**Reply by david-zwicker** [reply]
Yeah, sounds as if Spyder is using a different package, which can unfortunately easily happen with the way packages are managed by python. Good luck with sorting this out :)
